internal/pkg/orchestrator/service.go

```diff
@@ -0,0 +1,209 @@
+type ServiceOpt interface {
+	apply(Service) Service
+}
+
+const (
+	// envTargetExeKey is the key for the environment variable value pointing to the
+	// target binary to instrument.
+	envTargetExeKey = "OTEL_GO_AUTO_TARGET_EXE"
+	// envServiceName is the key for the envoriment variable value containing the service name.
+	envServiceNameKey = "OTEL_SERVICE_NAME"
+	// envResourceAttrKey is the key for the environment variable value containing
+	// OpenTelemetry Resource attributes.
+	envResourceAttrKey = "OTEL_RESOURCE_ATTRIBUTES"
+	// serviceNameDefault is the default service name prefix used if a user does not provide one.
+	serviceNameDefault = "unknown_service"
+)
+
+func (s Service) applyEnv() Service {
+	if v, ok := os.LookupEnv(envTargetExeKey); ok {
+		s.exePath = v
+	}
+	if v, ok := os.LookupEnv(envServiceNameKey); ok {
+		s.serviceName = v
+	} else {
+		s = s.applyResourceAtrrEnv()
+		if s.serviceName == "" {
+			s = s.setDefualtServiceName()
+		}
+	}
+	return s
+}
+
+func (s Service) setDefualtServiceName() Service {
+	if s.exePath != "" {
+		s.serviceName = fmt.Sprintf("%s:%s", serviceNameDefault, filepath.Base(s.exePath))
+	} else {
+		s.serviceName = serviceNameDefault
+	}
+	return s
+}
+
+func (s Service) applyResourceAtrrEnv() Service {
+	attrs := strings.TrimSpace(os.Getenv(envResourceAttrKey))
+	serviceName := serviceNameFromAttrs(attrs)
+	if serviceName != "" {
+		s.serviceName = serviceName
+	}
+	return s
+}
+
+func serviceNameFromAttrs(attrs string) string {
+	serviceName := ""
+	if attrs == "" {
+		return serviceName
+	}
+
+	pairs := strings.Split(attrs, ",")
+	for _, p := range pairs {
+		k, v, found := strings.Cut(p, "=")
+		if !found {
+			continue
+		}
+		key := strings.TrimSpace(k)
+		if key == string(semconv.ServiceNameKey) {
+			serviceName = strings.TrimSpace(v)
+		}
+	}
+	return serviceName
+}
+
+// Validate validates [Service] and returns an error if not valid.
+func (s Service) Validate() error {
+	if s.pid != 0 {
+		return validatePID(s.pid)
+	}
+
+	if s.monitorAll {
+		return nil
+	}
+	if s.serviceName == "" {
+		return fmt.Errorf("serviceName is nil")
+	}
+
+	if s.exePath == "" {
+		return fmt.Errorf("execPath is nil")
+	}
+
+	return nil
+}
+
+func validatePID(pid int) error {
+	p, err := os.FindProcess(pid)
+	if err != nil {
+		return fmt.Errorf("can't find process with pid %d", pid)
+	}
+	err = p.Signal(syscall.Signal(0))
+	if err != nil {
+		return fmt.Errorf("process with pid %d does not exist", pid)
+	}
+	return nil
+}
+
+type fnOpt func(Service) Service
+
+func (o fnOpt) apply(c Service) Service { return o(c) }
+
+// WithTarget returns an [InstrumentationOption] defining the target binary for
+// [Instrumentation] that is being executed at the provided path.
+//
+// If multiple of these options are provided to an [Instrumentation], the last
+// one will be used.
+//
+// If OTEL_GO_AUTO_TARGET_EXE is defined it will take precedence over any value
+// passed here.
+func WithTarget(path string) ServiceOpt {
```

Comment:
Is there a way to avoid this duplication with the functions in `instrumentation.go` ?

Reply:
I would like remove `instrumentation.go` . I will do so in future PRs. After orchestrator implementation, there is no need for `instrumentation.go` . I didn't want to introduce breaking changes in this PR.